go-controller/pkg/clustermanager/clustermanager.go

```diff
@@ -0,0 +1,271 @@
+func (cm *ClusterManager) Start(nodeName string, ctx context.Context) error {
+	klog.Infof("Cluster manager Started")
+	metrics.RegisterClusterManagerBase()
+
+	// Set up leader election process first.
+	// User lease resource lock as configmap and endpoint lock support is removed from leader election library.
+	rl, err := resourcelock.New(
+		resourcelock.LeasesResourceLock,
+		config.Kubernetes.OVNConfigNamespace,
+		"ovn-kubernetes-cluster-manager",
+		cm.client.CoreV1(),
+		cm.client.CoordinationV1(),
+		resourcelock.ResourceLockConfig{
+			Identity:      nodeName,
+			EventRecorder: cm.recorder,
+		},
+	)
+	if err != nil {
+		return err
+	}
+
+	lec := leaderelection.LeaderElectionConfig{
+		Lock:            rl,
+		LeaseDuration:   time.Duration(config.ClusterMgrHA.ElectionLeaseDuration) * time.Second,
+		RenewDeadline:   time.Duration(config.ClusterMgrHA.ElectionRenewDeadline) * time.Second,
+		RetryPeriod:     time.Duration(config.ClusterMgrHA.ElectionRetryPeriod) * time.Second,
+		ReleaseOnCancel: true,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(ctx context.Context) {
+				klog.Infof("Won leader election; in active mode")
+				// run the cluster controller to init the cluster manager
+				start := time.Now()
+				defer func() {
+					end := time.Since(start)
+					metrics.MetricClusterManagerReadyDuration.Set(end.Seconds())
+				}()
+
+				// run only on the active master node.
+				if err := cm.Run(); err != nil {
+					panic(err.Error())
+				}
+			},
+			OnStoppedLeading: func() {
+				// This node was leader and it lost the election.
+				// Whenever the node transitions from leader to follower,
+				// we need to handle the transition properly like clearing
+				// the cache. It is better to exit for now.
+				// kube will restart and this will become a follower.
+				klog.Infof("No longer leader; exiting")
+				os.Exit(0)
+			},
+			OnNewLeader: func(newLeaderName string) {
+				if newLeaderName != nodeName {
+					klog.Infof("Lost the election to %s; in standby mode", newLeaderName)
+				}
+			},
+		},
+	}
+
+	leaderelection.SetProvider(ovnkubeClusterManagerLeaderMetricsProvider{})
+	leaderElector, err := leaderelection.NewLeaderElector(lec)
+	if err != nil {
+		return err
+	}
+
+	cm.wg.Add(1)
+	go func() {
+		leaderElector.Run(ctx)
+		klog.Infof("Stopped leader election")
+		cm.wg.Done()
+	}()
+
+	return nil
+}
+
+func (cm *ClusterManager) Stop() {
+	close(cm.stopChan)
+	cm.defaultNetClusterController.Stop()
+
+	if cm.secondaryNetClusterControllerManager != nil {
+		cm.secondaryNetClusterControllerManager.Stop()
+	}
+}
+
+// Run starts the watch factory, inits and runs the default network cluster controller
+func (cm *ClusterManager) Run() error {
+	metrics.RegisterClusterManagerFunctional()
+	if err := cm.defaultNetClusterController.Init(); err != nil {
+		return err
+	}
+
+	if cm.secondaryNetClusterControllerManager != nil {
```

Comment:
nit: I think its clearer if you used:
```
if config.OVNKubernetesFeature.EnableMultiNetwork {
```

Reply:
I can change it.

I thought about it and from my 'C' experience this sounded correct.

Eg.

char *p = NULL;

if (some_flag_set) {
    p = malloc(...)
}

And later point it feels more safer to check if p is not NULL before accessing it.

i.e

if (p) {
    do_what_ever_to_p
}

instead of

if (some_flag_set) {
   do_what_ever_to_p
}


If I've not convinced still please let me know I'll change it.

Please do +1 on this comment to change it.